dev/camdenhells/REQUIREMENTS.md

```diff
@@ -324,41 +350,75 @@ This does not preclude performance impacts; clearly there will be limits, but ha
   * Isolated - from all (other) render thread/event processing
     * Frame draws, window re-sizes, queued events, other CRUD transactions, etc.
     * Implies running on the render thread as an event, queued from any thread
-    * TODO how to handle queued transactions after window detach/loss of context...
+    * Queued transactions after window detach/loss of context are dropped
   * Durable - process persistence scope
     * Combined state preserved across Activities and render contexts
     * Essentially a requirement to provide automatic restores
 3. Provide "successful completion" callback
-  * Provide a variant that includes queueing to main/UI thread ? 
-
-# 4. Testing, Debugging, and Reliability Requirements
-
-1. Automated unit/integration test harness and tests 
-  * Details ?
-2. Internal runtime logging/tracing/assertions?
-  * E.g. for integration testing and for debugging support issues
-3. Compile-time API-usage validation, e.g. annotations ?
-  * Are these enforceable e.g. via lint?
-4. Runtime API-usage validation for the app, e.g. debug-only assertions/exceptions?
-  * 
-5. Debug visualizations (#1436)
-  * Trident, camera eyeball/frustum, lighting rays/lines, object bounds
-  * Are there (non-debug) app use cases for some of these as well? e.g.: 
-    * Interactive camera/lighting selections from a global-view PIP inset 
-    * Theater/stage lighting design apps
-    * Interior design apps - door/window/furniture/light placements
-  * If so perhaps just call them e.g. camera/light/bounds visualizations  
+  * Provide a variant that includes queueing to main/UI thread (see above)
+
+# 4. Testing, Validation, and Debugging Requirements
+
+## 4.1 Compile-time
+
+  * Consistently specify any constraints on all member and parameter values
+    * Specify all references as either nullable or non-null
+    * Specify integer and float ranges whenever less than built-in language limits
+  * ~Validate public API-usage [2.1]~
+     * ~Create an Android lint plugin to detect invocations of non API methods?~
+     * Seems like a reliable library-specific tool would be a lot of work/maintenance, and unlikely that client developers would download/install/use it anyway
+
+## 4.2 Build-time 
+
+  Help! I'm trying to define the line on what should be unit tested and what should not, and don't even understand the hardware/software environment in which unit tests run (i.e. emulators use what version of OpenGL? is there an Android process/activity context? etc)
```

Comment:
There are two kinds of unit tests. Those that exist in a java context only (such as the tests for the math package - Matrix4, etc.), and those that exist in a GL context (such as VBO tests). At present, the GL tests are setup around using an Android activity and surfaceview, but this may change if we create our own GL thread. The tests only need a valid GL context, they don't need to display. 

At the moment, the travis build can only execute the non-GL unit tests due to a limitation of the emulator combined with travis. Essentially because it is a headless emulator, GL support is not included. Hardware tests are currently executed on my private device farm, with occasional runs by @ToxicBakery to supplement the available devices. Additional users would be welcomed. We can add any of these devices to the list in the `v2.0-development` `README.md`.

As for what should be unit tested, anything that can be black boxed. Essentially, if you can expect a certain output for a given input because it interacts only with itself, then it is a candidate for unit testing. Minimal interaction is a candidate for integration testing - which I have been adding as well. Scene graphs are a good example of this.

Reply:
Thanks, that helps a lot.  Will update accordingly.